raft/src/main/java/org/apache/kafka/raft/internals/PartitionListener.java

```diff
@@ -0,0 +1,259 @@
+     * @param bufferSupplier the supplier of byte buffers
+     * @param maxBatchSizeBytes the maximum size of record batch
+     * @param logContext the log context
+     */
+    public PartitionListener(
+        Optional<VoterSet> staticVoterSet,
+        ReplicatedLog log,
+        RecordSerde<?> serde,
+        BufferSupplier bufferSupplier,
+        int maxBatchSizeBytes,
+        LogContext logContext
+    ) {
+        this.log = log;
+        this.voterSetHistory = new VoterSetHistory(staticVoterSet);
+        this.serde = serde;
+        this.bufferSupplier = bufferSupplier;
+        this.maxBatchSizeBytes = maxBatchSizeBytes;
+        this.logger = logContext.logger(this.getClass());
+    }
+
+    /**
+     * Must be called whenever the {@code log} has changed.
+     */
+    public void updateListener() {
+        maybeLoadSnapshot();
+        maybeLoadLog();
+    }
+
+    /**
+     * Remove the head of the log until the given offset.
+     *
+     * @param endOffset the end offset (exclusive)
+     */
+    public void truncateTo(long endOffset) {
+        synchronized (voterSetHistory) {
+            voterSetHistory.truncateTo(endOffset);
+        }
+        synchronized (kraftVersionHistory) {
+            kraftVersionHistory.truncateTo(endOffset);
+        }
+    }
+
+    /**
+     * Remove the tail of the log until the given offset.
+     *
+     * @param @startOffset the start offset (inclusive)
+     */
+    public void trimPrefixTo(long startOffset) {
+        synchronized (voterSetHistory) {
+            voterSetHistory.trimPrefixTo(startOffset);
+        }
+        synchronized (kraftVersionHistory) {
+            kraftVersionHistory.trimPrefixTo(startOffset);
+        }
+    }
+
+    /**
+     * Returns the last voter set.
+     */
+    public VoterSet lastVoterSet() {
+        synchronized (voterSetHistory) {
+            return voterSetHistory.lastValue();
+        }
+    }
+
+    /**
+     * Rturns the voter set at a given offset.
+     *
+     * @param offset the offset (inclusive)
+     * @return the voter set if one exist, otherwise {@code Optional.empty()}
+     */
+    public Optional<VoterSet> voterSetAtOffset(long offset) {
+        long fixedNextOffset = nextOffset;
+        if (offset >= fixedNextOffset) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Attempting the read the voter set at an offset (%d) which kraft hasn't seen (%d)",
+                    offset,
+                    fixedNextOffset - 1
+                )
+            );
+        }
+
+        synchronized (voterSetHistory) {
+            return voterSetHistory.valueAtOrBefore(offset);
+        }
+    }
+
+    /**
+     * Returns the finalized kraft version at a given offset.
+     *
+     * @param offset the offset (inclusive)
+     * @return the finalized kraft version if one exist, otherwise 0
+     */
+    public short kraftVersionAtOffset(long offset) {
```

Comment:
nit: it might be possible to use a (private) generic method to avoid the duplication

Reply:
What do you mean? I created `checkOffsetIsValid` is that what you meant?